src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,1051 @@
+  address small_hs_jmp_table[10];   // Jump table for small haystacks
+  int jmp_ndx = 0;
+  __ align(CodeEntryAlignment);
+  address start = __ pc();
+  __ enter();  // required for proper stackwalking of RuntimeStub frame
+
+  const Register haystack     = rdi;
+  const Register haystack_len = rsi;
+  const Register needle       = rdx;
+  const Register needle_len   = rcx;
+
+  const Register save_ndl_len = r12;
+
+  const XMMRegister save_r12  = xmm4;
+  const XMMRegister save_r13  = xmm5;
+  const XMMRegister save_r14  = xmm6;
+  const XMMRegister save_r15  = xmm7;
+  const XMMRegister save_rbx  = xmm8;
+  const XMMRegister save_rsi  = xmm9;
+  const XMMRegister save_rdi  = xmm10;
+  const XMMRegister save_rcx  = xmm11;
+  const XMMRegister save_r8   = xmm12;
+  const XMMRegister save_r9   = xmm13;
+
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  //                         AVX2 code
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  if (VM_Version::supports_avx2()) {  // AVX2 version
+    Label L_begin;
+
+    Label L_returnRBP, L_returnRAX, L_checkRangeAndReturn;
+    Label L_bigCaseFixupAndReturn, L_small7_8_fixup, L_checkRangeAndReturnRCX;
+    Label L_returnZero, L_haystackGreaterThan31, L_copyToStackDone, L_bigSwitchTop, L_copyToStack, L_smallSwitchTop;
+    Label L_bigCaseDefault, L_smallCaseDefault;
+
+    address jump_table;
+    address jump_table_1;
+
+    // Jump past jump table setups to get addresses of cases.
+    __ jmp(L_begin);
+
+    ////////////////////////////////////////////////////////////////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////////////
+    //
+    // Set up jump table entries for both small and large haystack switches.
+
+// Small case 1:
+    // Unroll for speed
+    small_hs_jmp_table[0] = __ pc();
+    {
+      Label L_tmp1, L_tmp2, L_tmp3, L_tmp4, L_tmp5, L_returnRDI, L_tail;
+
+      __ movzbl(rcx, Address(r14, 0));
+      __ movl(rax, rbx);
+      __ andl(rax, 0xf);
+      __ movl(rdx, 0x10);
+      __ subq(rdx, rax);
+      __ cmpq(rdx, rsi);
+      __ movq(rax, rsi);
+      __ cmovl(Assembler::below, rax, rdx);
+      __ testq(rax, rax);
+      __ je_b(L_tmp1);
+      __ xorl(rdi, rdi);
+
+      __ bind(L_tmp2);
+      __ cmpb(Address(rbx, rdi, Address::times_1), rcx);
+      __ je(L_returnRDI);
+      __ incq(rdi);
+      __ cmpq(rax, rdi);
+      __ jne(L_tmp2);
+
+      __ bind(L_tmp1);
+      __ cmpq(rdx, rsi);
+      __ jae(L_returnRBP);
+      __ movq(rdi, rsi);
+      __ subq(rdi, rax);
+      __ movq(rdx, rdi);
+      __ andq(rdx, -16);
```

Comment:
Hi @asgibbons , may I request you to please use meaningful names instead of directly using actual GRP names.